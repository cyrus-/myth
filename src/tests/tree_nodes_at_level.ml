let s = "
type bool =
  | True
  | False

type tree =
  | Leaf
  | Node of tree * bool * tree

type nat =
  | O
  | S of nat

let rec sum (n1:nat) (n2:nat) : nat =
      match n1 with
      | O -> n2
      | S (n1p) -> S (sum n1p n2)
;;

let tree_nodes_at_level : tree -> nat -> nat |>
  { Leaf =>
    ( 0 => 0
    | 1 => 0
    | 2 => 0
    | 3 => 0
    )
  | Node (Leaf, True, Leaf) =>
    ( 0 => 1
    | 1 => 0
    | 2 => 0
    | 3 => 0
    )
  | Node (Node (Leaf, True, Leaf), True, Leaf) =>
    ( 0 => 1
    | 1 => 1
    | 2 => 0
    | 3 => 0
    )
  | Node (Leaf, True, Node (Leaf, True, Leaf)) =>
    ( 0 => 1
    | 1 => 1
    | 2 => 0
    | 3 => 0
    )
  | Node (Node (Leaf, True, Leaf), True, Node (Leaf, True, Leaf)) =>
    ( 0 => 1
    | 1 => 2
    | 2 => 0
    | 3 => 0
    )
  | Node (Node (Node(Leaf, True, Leaf), True, Node (Leaf, True, Leaf)), True, Leaf) =>
    ( 0 => 1
    | 1 => 1
    | 2 => 2
    | 3 => 0)
  } = ?"
