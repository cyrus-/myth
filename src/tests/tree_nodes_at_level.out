(* Generated Thu Nov 13 00:35:56 2014 *)
let tree_nodes_at_level : tree -> nat -> nat =
  let rec f1 (t1:tree) : nat -> nat =
    fun (n1:nat) ->
      match t1 with
        | Leaf -> O
        | Node (t2, b1, t3) -> (match n1 with
                                  | O -> S (O)
                                  | S (n2) -> sum (f1 t3 n2) (f1 t2 n2))
  in
    f1
;;
